src/npcmove.cpp

```diff
@@ -2867,50 +2869,23 @@ void npc::set_destination()
     if( needs.empty() ) { // We don't need anything in particular.
         needs.push_back( need_none );
     }
-    std::vector<std::string> options;
-    switch( needs[0] ) {
-        case need_ammo:
-            options.push_back( "house" );
-        /* fallthrough */
-        case need_gun:
-            options.push_back( "s_gun" );
-            break;
-
-        case need_weapon:
-            options.push_back( "s_gun" );
-            options.push_back( "s_sports" );
-            options.push_back( "s_hardware" );
-            break;
-
-        case need_drink:
-            options.push_back( "s_gas" );
-            options.push_back( "s_pharm" );
-            options.push_back( "s_liquor" );
-        /* fallthrough */
-        case need_food:
-            options.push_back( "s_grocery" );
-            break;
-
-        default:
-            options.push_back( "house" );
-            options.push_back( "s_gas" );
-            options.push_back( "s_pharm" );
-            options.push_back( "s_hardware" );
-            options.push_back( "s_sports" );
-            options.push_back( "s_liquor" );
-            options.push_back( "s_gun" );
-            options.push_back( "s_library" );
-    }
-
-    const std::string dest_type = random_entry( options );
 
     // We need that, otherwise find_closest won't work properly
     // TODO: Allow finding sewers and stuff
     tripoint surface_omt_loc = global_omt_location();
     surface_omt_loc.z = 0;
 
-    goal = overmap_buffer.find_closest( surface_omt_loc, dest_type, 0, false );
-    add_msg( m_debug, "New goal: %s at %d,%d,%d", dest_type.c_str(), goal.x, goal.y, goal.z );
+    std::string dest_type = overmap_locations::get_random_terrain( need_id( needs[ 0 ] ) );
+    if( dest_type.empty() ) {
+        dest_type = "field";
```

Comment:
This looks like a hack. Normally, `dest_type` shouldn't return an empty string (BTW it souldn't return any `string` whatsoever, there's an ad hoc identifier type for this purpose).

Reply:
Yes, kind of a hack for city-less worlds (see #22270), so it would be possible to use two different functions `find_random` and `find_closest`, otherwise NPC would probably stuck on single overmap tile in city-less worlds.